hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java

```diff
@@ -161,31 +159,7 @@ private void scan() {
                 Long callId = entry.getKey();
                 Invocation invocation = entry.getValue();
 
-                /*
-                * The reason for the following if check is a workaround for the problem explained below.
-                *
-                * Problematic scenario :
-                * If an invocation with callId 1 is retried twice for any reason,
-                * two new innovations created and registered to invocation registry with callId’s 2 and 3 respectively.
-                * Both new invocations are sharing the same operation
-                * When one of the new invocations, say the one with callId 2 finishes, it de-registers itself from the
-                * invocation registry.
-                * When doing the de-registration it sets the shared operation’s callId to 0.
-                * After that when the invocation with the callId 3 completes, it tries to de-register itself from
-                * invocation registry
-                * but fails to do so since the invocation callId and the callId on the operation is not matching anymore
-                * When InvocationMonitor thread kicks in, it sees that there is an invocation in the registry,
-                * and asks whether invocation is finished or not.
-                * Even if the remote node replies with invocation is timed out,
-                * It can’t be de-registered from the registry because of aforementioned non-matching callId scenario.
-                *
-                * Workaround:
-                * When InvocationMonitor kicks in, it will do a check for invocations that are completed
-                * but their callId's are not matching with their operations. If any invocation found for that type,
-                * it is removed from the invocation registry.
-                *
-                * */
-                if (!callIdMatches(callId, invocation) && isDone(invocation)) {
+                if (duplicate(invocation, callId, iterator)) {
                     iterator.remove();
```

Comment:
There are two `iterator.remove` in case of `duplicate` returns `true`. One is here and other is inside `duplicate`. One of them should be removed.